### PR TITLE
Transaction builders, TokenPayment, transfer & execute

### PR DESCRIPTION
**Pull request by andreibancioiu**
Following (partly) erdjs:

 - Implement transactions builders (for contracts, for transfers).
 - Implement transfer & execute.
 - Implement `TokenPayment` (as in erdjs). 
 - Replace `Account` by `AccountNonceHolder`.
 - Add `compute_contract_address()`.
 - Adjust type `ITransactionValue`.

### CODE REVIEW

File: erdpy_core/address.py [hunk 1]
```diff
@@ -103,3 +107,45 @@ def _decode_bech32(value: str) -> Tuple[str, bytes]:
...
+        return True
+
+    zero_address = bytearray(32)
+    if pubkey == zero_address:
+        return True
+
+    return False
+
+
+def compute_contract_address(deployer: IAddress, deployment_nonce: INonce, address_hrp: str) -> Address:
```
Comment (andreibancioiu) [inline erdpy_core/address.py:142]: Copied from erdpy.

File: erdpy_core/token_payment.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
+from decimal import ROUND_DOWN, Context, Decimal, localcontext
+from typing import Union
+
+from erdpy_core.constants import EGLD_NUM_DECIMALS, EGLD_TOKEN_IDENTIFIER
+from erdpy_core.interfaces import INonce, ITokenIdentifier
+
+
+class TokenPayment:
```
Comment (andreibancioiu) [inline erdpy_core/token_payment.py:8]: Resembles the erdjs implementation.

File: erdpy_core/transaction_builders/relayed_builders.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class RelayedTransactionV1Builder:
```
Comment (andreibancioiu) [inline erdpy_core/transaction_builders/relayed_builders.py:1]: Left for future PRs.

File: pyproject.toml [hunk 4]
```diff
@@ -15,6 +15,9 @@ classifiers = [
     "License :: OSI Approved :: MIT License",
     "Operating System :: OS Independent",
 ]
+dependencies = [
+  "pycryptodomex==3.16.0"
```
Comment (andreibancioiu) [inline pyproject.toml:19]: Necessary for `keccak` (to compute the address of a smart contract).

**Review by schimih — APPROVED** [review]
nice! :+1:

**Review by schimih — APPROVED** [review]

File: erdpy_core/transaction_builders/base_builder.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+
+class IBaseConfiguration(Protocol):
+    chain_id: IChainID
+    min_gas_price: IGasPrice
+    min_gas_limit: IGasLimit
+    gas_limit_per_byte: IGasLimit
+
+
+class BaseBuilder:
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply erdpy_core/transaction_builders/base_builder.py:20]: Renamed.

File: erdpy_core/transaction_builders/base_builder.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        self.gas_limit_per_byte = config.gas_limit_per_byte
+
+        self.nonce = nonce
+        self.value = value
+        self.gas_limit = gas_limit
+        self.gas_price = gas_price or config.min_gas_price
+        self.sender: Optional[IAddress] = None
+        self.receiver: Optional[IAddress] = None
+
+    def build_transaction(self) -> Transaction:
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply erdpy_core/transaction_builders/base_builder.py:39]: Renamed.

File: erdpy_core/transaction_builders/esdt_builders.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+                 num_decimals: int,
+                 can_freeze: bool = False,
+                 can_wipe: bool = False,
+                 can_pause: bool = False,
+                 can_mint: bool = False,
+                 can_burn: bool = False,
+                 can_change_owner: bool = False,
+                 can_upgrade: bool = False,
+                 can_add_special_roles: bool = False,
+                 nonce: Optional[INonce] = None,
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply erdpy_core/transaction_builders/esdt_builders.py:32]: Yes, some users might choose to provide it at an early time (in the builder constructor), while others would only know the nonce at a later time, shortly prior signing the transaction.

**Review by schimih — APPROVED** [review]

**Review by claudiu725 — DISMISSED** [review]

**Review by schimih — DISMISSED** [review]

**Review by schimih — APPROVED** [review]

**Review by claudiu725 — APPROVED** [review]

**Pull request merged**